compiler/sem/semexprs.nim

```diff
@@ -1605,21 +1605,72 @@ proc buildOverloadedSubscripts(n: PNode, ident: PIdent): PNode =
   for s in n: result.add s
 
 proc semDeref(c: PContext, n: PNode): PNode =
-  checkSonsLen(n, 1, c.config)
-  n[0] = semExprWithType(c, n[0])
-  let a = getConstExpr(c.module, n[0], c.idgen, c.graph)
-  if a != nil:
-    if a.kind == nkNilLit:
-      localReport(c.config, n, reportSem rsemDisallowedNilDeref)
-    n[0] = a
-  result = n
-  var t = skipTypes(n[0].typ, {tyGenericInst, tyVar, tyLent, tyAlias, tySink, tyOwned})
-  case t.kind:
-    of tyRef, tyPtr:
-      n.typ = t.lastSon
+  ## analyse deref such as `[]foo`, when given an nkBracketExpr converts it to
+  ## an nkDerefExpr and evals it, if given an nkDerefExpr evals it, nkError is
+  ## passed through.
+  ##
+  ## it to a  behaves differently depending upon whether
```

Comment:
Speaking of "simple English" just mentioned in the `#main` - maybe it could be rewritten as "There are two different cases: when target constant expression (static) and when it is not (dynamic)".

Can leave it as it is now, I just noticed it might be a good example of "let's write in a simplified English".

Reply:
rewritten.